### DISCUSSION TITLE
Looking for something similar to Clump

### DISCUSSION THREAD

**Discussion by aboisvert** [lampepfl/gears#104] [General]
Hi,

I'm using Gears -- mostly Future/await at the moment -- and happy to far.   However, as our application grows, and we look towards optimizing some flows, we are doing more and more work around batching requests as a means of reducing latency and improving efficiency.

There are a few libraries that provide helpful tools for batching requests, in particular,
- Clump (https://github.com/getclump/clump)
- Parseq + Parseq Batching (https://github.com/linkedin/parseq/tree/master/subprojects/parseq-batching)

My questions are:
- are there any similar existing tools for Gears? (built-in or 3rd-party)
- has the design space for async request batching been explored with Gears?
- specifically, would it be possible to port Clump to a direct-style API?  Or does something like Clump necessitate some sort of monadic/applicative wrapper?

thanks!

**Comment by He-Pin** [comment]
I think you can build something on top the current future/promise with a timer for batching, eg the groupWeithedWithin in pekko/akka stream

**Reply by aboisvert** [reply]
Good thought.  I think could be a relatively easy solution but I'm not too keen on a timer-based approach because this is the context of a web application where we optimize for response time first, and any timer-based solution would deviate from this goal.  

Solutions like Clump/Parseq end up building a DAG of data dependencies and will trigger batching at strategic points, avoiding delays induced by timers, but also yielding predictable batching behavior.  This is a bit more what I'm looking for.
